### ISSUE TITLE
Investigate the possibility of using `jax` as backend (ongoing)

### ISSUE THREAD

**Issue by james77777778** [james77777778/keras-aug#105]
Can KerasAug uses jax as its backend?

Refs:
- https://github.com/deepmind/deepmind-research/tree/master/byol
- https://github.com/google-research/scenic/blob/main/scenic/projects/baselines/detr/input_pipeline_detection.py
- https://github.com/google-research/scenic/blob/main/scenic/projects/baselines/detr/transforms.py
- https://github.com/deepmind/dm_pix

Pseudo code:

```python
class BaseLayer(nn.Module):
    ...

    @nn.compact
    def __call__(self, inputs):
        self.get_random_transformation_batch()
        self.augment_images()
        self.augment_bounding_boxes()
        self.augment_segmentation_masks()
        self.augment_keypoints()

        return outputs
```
- Support dense array first

**Comment by IMvision12** [comment]
@james77777778 I guess you have explored many repo's so is it possible to use Jax as the backend?

**Comment by james77777778** [comment]
@IMvision12 
Actually, I'm still learning the basics of `jax` and `flax` and try to figure out what is the best practice to implement the preprocessing / augmentation layers using them.

At first glance, I might place the jax/flax-based layers at a new folder `keras_aug/jax` and keep original implementations untouched.

**Comment by james77777778** [comment]
```python
import cv2
import flax.linen as nn
import jax
import jax.numpy as jnp
import matplotlib.pyplot as plt
import numpy as np


class RandomFlip(nn.Module):
    mode: str = "horizontal"
    rate: float = 0.5

    @nn.compact
    def __call__(self, inputs):
        images = inputs.get("images", None)
        rng_h = self.make_rng("rng")
        flip_horizontals = jax.random.uniform(rng_h, shape=(images.shape[0], 1))
        rng_v = self.make_rng("rng")
        flip_verticals = jax.random.uniform(rng_v, shape=(images.shape[0], 1))
        flip_horizontals = flip_horizontals[:, jnp.newaxis, jnp.newaxis, :]
        flip_verticals = flip_verticals[:, jnp.newaxis, jnp.newaxis, :]
        if "horizontal" in self.mode:
            images = jnp.where(
                flip_horizontals > (1.0 - self.rate),
                jnp.flip(images, axis=2),
                images,
            )
        if "vertical" in self.mode:
            images = jnp.where(
                flip_verticals > (1.0 - self.rate),
                jnp.flip(images, axis=1),
                images,
            )
        inputs["images"] = images
        return inputs


image = cv2.imread("image.jpg")
print(image.shape)
inputs = {"images": jnp.array(np.expand_dims(image, axis=0))}

random_flip = RandomFlip("horizontal_and_vertical")
init_rngs = {"rng": jax.random.PRNGKey(0)}
variables = random_flip.init(
    init_rngs, {"images": jnp.empty((1, *image.shape))}
)

next_rngs = {"rng": jax.random.PRNGKey(9)}
outputs = random_flip.apply(variables, inputs)

plt.imshow(np.array(outputs["images"][0, ..., ::-1]))
plt.show()
plt.close()

```

an example to implement RandomFlip with jax and flax.

**Comment by IMvision12** [comment]
That's awesome, I will add some layers implemented in Jax/flax, I am also still learning!!!

**Comment by james77777778** [comment]
Closed by #127
